lib/integrations/awsoidc/deployservice_iam_config_test.go

```diff
@@ -216,35 +222,105 @@ func TestDeployServiceIAMConfig(t *testing.T) {
+	if golden.ShouldSet() {
+		golden.Set(t, buf.Bytes())
+	}
+	require.Equal(t, string(golden.Get(t)), buf.String())
+}
+
 type mockDeployServiceIAMConfigClient struct {
 	CallerIdentityGetter
-	existingRoles []string
+	existingRoles map[string]mockRole
 }
 
 // CreateRole creates a new IAM Role.
 func (m *mockDeployServiceIAMConfigClient) CreateRole(ctx context.Context, params *iam.CreateRoleInput, optFns ...func(*iam.Options)) (*iam.CreateRoleOutput, error) {
-	alreadyExistsMessage := fmt.Sprintf("Role %q already exists.", *params.RoleName)
-	if slices.Contains(m.existingRoles, *params.RoleName) {
+	roleName := aws.ToString(params.RoleName)
+	alreadyExistsMessage := fmt.Sprintf("Role %q already exists.", roleName)
+	if _, found := m.existingRoles[roleName]; found {
 		return nil, &iamtypes.EntityAlreadyExistsException{
 			Message: &alreadyExistsMessage,
 		}
 	}
-	m.existingRoles = append(m.existingRoles, *params.RoleName)
+	m.existingRoles[roleName] = mockRole{
+		assumeRolePolicyDoc: params.AssumeRolePolicyDocument,
+		tags:                params.Tags,
+	}
 
-	return nil, nil
+	return &iam.CreateRoleOutput{}, nil
 }
 
 // PutRolePolicy creates or replaces a Policy by its name in a IAM Role.
 func (m *mockDeployServiceIAMConfigClient) PutRolePolicy(ctx context.Context, params *iam.PutRolePolicyInput, optFns ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error) {
-	if !slices.Contains(m.existingRoles, *params.RoleName) {
-		noSuchEntityMessage := fmt.Sprintf("Role %q does not exist.", *params.RoleName)
+	roleName := aws.ToString(params.RoleName)
+	if _, found := m.existingRoles[roleName]; !found {
+		noSuchEntityMessage := fmt.Sprintf("Role %q does not exist.", roleName)
 		return nil, &iamtypes.NoSuchEntityException{
 			Message: &noSuchEntityMessage,
 		}
 	}
 	return nil, nil
 }
 
+func (m *mockDeployServiceIAMConfigClient) GetRole(ctx context.Context, params *iam.GetRoleInput, _ ...func(*iam.Options)) (*iam.GetRoleOutput, error) {
+	roleName := aws.ToString(params.RoleName)
+	role, found := m.existingRoles[roleName]
+	if !found {
+		return nil, trace.NotFound("role %q not found", roleName)
+	}
+	return &iam.GetRoleOutput{
+		Role: &iamtypes.Role{
+			AssumeRolePolicyDocument: role.assumeRolePolicyDoc,
+		},
+	}, nil
+}
+
+func (m *mockDeployServiceIAMConfigClient) UpdateAssumeRolePolicy(ctx context.Context, params *iam.UpdateAssumeRolePolicyInput, optFns ...func(*iam.Options)) (*iam.UpdateAssumeRolePolicyOutput, error) {
+	roleName := aws.ToString(params.RoleName)
+	role, found := m.existingRoles[roleName]
+	if !found {
+		return nil, trace.NotFound("role not found")
+	}
+
+	role.assumeRolePolicyDoc = params.PolicyDocument
+	m.existingRoles[roleName] = role
+
+	return &iam.UpdateAssumeRolePolicyOutput{}, nil
+}
+
+func (m *mockDeployServiceIAMConfigClient) TagRole(ctx context.Context, params *iam.TagRoleInput, _ ...func(*iam.Options)) (*iam.TagRoleOutput, error) {
+	roleName := aws.ToString(params.RoleName)
```

Comment:
hm there's actually no test exercising the tagging and I dislike testing "X function was called and produced Y as a side-effect" in tests, but since it's already written I'll just leave as-is 😄